### DISCUSSION TITLE
How to DRY with column expressions

### DISCUSSION THREAD

**Discussion by seperman** [sqlalchemy/sqlalchemy#9826] [Usage Questions] [answered]
Hello, in the body of a declarative model I have expressions such as:

```py
_only_unique_unresolved_alerts = and_(~resolved, data_source_model_id == None)
```

These were then use for example in Postgres where clause:

```py
        Index('_uq', signature,
              unique=True,
              postgresql_where=_only_unique_unresolved_alerts
              ),
```

An other places in the same model.

Since upgrading to SQLAlchemy 2.0.13, I get a warning:

```

    def _warn_for_decl_attributes(self, cls, key, c):
        if isinstance(c, expression.ColumnClause):
            util.warn(
                "Attribute '%s' on class %s appears to be a non-schema "
                "'sqlalchemy.sql.column()' "
                "object; this won't be part of the declarative mapping" %
                (key, cls))
```

How can I stop getting these warnings? It is fine with me that this specific line is not part of the declarative mapping.

**Comment by zzzeek** [comment]
there's no model here so I can't reproduce but it appears you are using a `column()` in a mapping somehow.   Use `Column` instead or otherwise please send along actionable details what you're doing.

**Comment by seperman** [comment]
@zzzeek Thanks for the prompt response. It is not because of lower case column:

```py
class AlertItem(BaseInitializer, Base):
    __tablename__ = 'alert_items'

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    data_source_model_id = Column(BigInteger, ForeignKey('data_source_models.id', name='_fk_alert_item_data_source_model'), nullable=True)
    resolved = Column(Boolean, default=False, nullable=False, doc="Is the alert resolved.")
    _only_unique_unresolved_alerts = and_(~resolved, data_source_model_id == None)

    __table_args__ = (
        Index('_uq', signature,
              unique=True,
              postgresql_where=_only_unique_unresolved_alerts
              ),
    )

    # _only_unique_unresolved_alerts is also used in some of the class methods
```

**Reply by zzzeek** [reply]
Hi what is "signature" ?  it's not included here and I cannot run the example

**Comment by seperman** [comment]
@zzzeek Here is a simpler one. I need to define the `not_signature` attribute to reuse it in multiple places.

```py
class AlertItem2(Base):
    __tablename__ = 'alert_items2'
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    signature = Column(String(), default='', nullable=True, doc="The signature of the alert so we can count it.")
    not_signature = ~signature
```

**Reply by zzzeek** [reply]
OK so the actual code is not what you showed above which is part of why this is hard to follow, can we confirm this is the code:

```py
    def _warn_for_decl_attributes(
        self, cls: Type[Any], key: str, c: Any
    ) -> None:
        if isinstance(c, expression.ColumnElement):
            util.warn(
                f"Attribute '{key}' on class {cls} appears to "
                "be a non-schema SQLAlchemy expression "
                "object; this won't be part of the declarative mapping. "
                "To map arbitrary expressions, use ``column_property()`` "
                "or a similar function such as ``deferred()``, "
                "``query_expression()`` etc. "
            )

```

note it doesnt mention "column()" at all, and this is the warning:

```
 SAWarning: Attribute 'not_signature' on class <class '__main__.AlertItem2'> appears to be a non-schema SQLAlchemy expression object; this won't be part of the declarative mapping. To map arbitrary expressions, use ``column_property()`` or a similar function such as ``deferred()``, ``query_expression()`` etc. 
```


I really need accurate details and examples to understand what we're looking at, because I thought this was something totally different.

There's a tension here between your desire to have arbitrary SQL expressions in the body of your class, and the extremely common mistake of placing arbitrary SQL expressions in the class body that one expects are part of the mapping where our option here would be to silently ignore it.

The quickest way to silence this warning is to just do what the warning says to do, which is map it with column_property().  You can use deferred=True or deferred() so that you will never see it in any queries:

```py

from sqlalchemy.orm import column_property

class AlertItem2(Base):
    __tablename__ = "alert_items2"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    signature = Column(String)
    data_source_model_id = Column(Integer)

    _only_unique_unresolved_alerts = column_property(
        data_source_model_id == None, deferred=True
    )

    # _only_unique_unresolved_alerts = deferred(data_source_model_id == None)

    __table_args__ = (
        Index(
            "_uq",
            signature,
            postgresql_where=_only_unique_unresolved_alerts,
            unique=True,
        ),
    )

```

something that certainly **should** work, but does not, is to declare the type as a ``ClassVar[]``, since we scan these also in declarative:

```py
class AlertItem2(Base):
    __tablename__ = "alert_items2"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    signature = Column(String)
    data_source_model_id = Column(Integer)

    _only_unique_unresolved_alerts: ClassVar[Any] = data_source_model_id == None

    __table_args__ = (
        Index(
            "_uq",
            signature,
            postgresql_where=_only_unique_unresolved_alerts,
            unique=True,
        ),
    )
```

we can add an issue for that use case and also augment the warning with this approach.

**Reply by CaselIT** [reply]
I like the classvar approach, since that's already similar at what dataclass does to detect fields vs things that are just in the class, so there is that precedent.

**Reply by seperman** [reply]
I tried the `column_property`. As a part of DRY, the `_only_unique_unresolved_alerts` was being used in the `index_where` parameter of `on_conflict_do_update` in `sqlalchemy.dialects.postgresql.insert`.
The attributes marked by `column_property` do not seem to be working with  `on_conflict_do_update` 's`index_where` clause. I've not dived too deep into it yet to give you a reproducible code. Just commenting that `column_property` caused errors.

**Reply by seperman** [reply]
For now I'm resorting to monkey patching until a better solution is found.

```py
from sqlalchemy.orm.decl_base import _ClassScanMapperConfig

_ClassScanMapperConfig._warn_for_decl_attributes = lambda *args, **kwargs: None
```

**Reply by zzzeek** [reply, marked as answer]
OK, certainly don't monkeypatch like that, just use [declared_attr](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.declared_attr) which is the traditional way to do this kind of thing, this uses plain `@classmethod` but you can also use a class property (slightly custom descriptor construct):

```py
from sqlalchemy import BigInteger
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Index
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import declared_attr
Base = declarative_base()

class AlertItem2(Base):
    __tablename__ = "alert_items2"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    signature = Column(String)
    data_source_model_id = Column(Integer)

    @classmethod
    def _only_unique_unresolved_alerts(cls):
        return cls.data_source_model_id == None

    @declared_attr
    def __table_args__(cls):
        return (
            Index(
            "_uq",
            cls.signature,
            postgresql_where=cls._only_unique_unresolved_alerts(),
            unique=True,
        ),
    )

e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)
```

**Reply by seperman** [reply]
Thank you @zzzeek and @CaselIT for prompt responses!

**Comment by CaselIT** [comment]
I think there currently isn't a way of suppressing that warning.
I also raised this on the review and the idea was "let's see the use case so an api for this can be defined". ref: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4526/comment/19dd4ec7_affd9f4e/
